### ISSUE TITLE
ceylon-model project

### ISSUE THREAD

**Issue by gavinking** [ceylon/ceylon-compiler#1763]
Currently, to run a Ceylon application that uses reified generics or the metamodel on the JVM, the following runtime dependencies are required:
- ceylon.language-1.1.0.car
- com.redhat.ceylon.typechecker-1.1.0.jar 
- com.redhat.ceylon.compiler.java-1.1.0.jar
- com.redhat.ceylon.common-1.1.0.jar
- org.jboss.modules-1.3.3.Final.jar
- com.redhat.ceylon.module-resolver-1.1.0.jar

On some platforms (Android), this is totally unacceptable.

We need to create a `com.redhat.ceylon.model.jar` that contains:
- the model package that currently resides in the typechecker, and
- the `AbstractModelLoader` and `ReflectionModelLoader`, along with just what they _absolutely_ need to function in an environment without `ceylon-runtime`. 

It would be acceptable if this new lightweight model loader is "incomplete" is certain ways. That is, it's OK if it doesn't support some modularity-related functionality, for example, package scanning, since it will likely run in a non-modular runtime. But it does need to provide support for basic reified-generics-based `is` tests, and basic metamodel usage, including reading annotations.

It is highly desirable that `com.redhat.ceylon.model.jar` has _no_ dependencies.

I would say that this is a critical issue for the whole platform.

**Comment by FroMage** [comment]
So I started to work on this thinking it would be straightforward, but after two days on it:

- the typechecker model is easy to extract
- the `ModuleManager` much harder as it pull dependencies on the `Tree` and is meant for typechecking more than just managing modules, so I have to split it in two and redesign it
- the compiler lazy model has grown back ties to the rest of the compiler, including `Naming`, `Decl` and things to do with annotations that I have to extract
- the extracted code still depends on `ceylon-common` (various things like `Versions`) which depends on the markdown renderer
- both models depend on some bits of the CMR which I can extract `ArtifactResult`, `JDKUtils` and others like `Repository` which are trickier
- the language module currently incorporates the runtime APIs like `Main` and `CeylonToolProvider`, which really depend on the compilers, so if we want the language module to not depend on them we have to extract those APIs yet again somewhere else
- extracting `Main` or the runner part of `CeylonToolProvider` sorta defeats the purpose that it's the entry point for running Ceylon programs in a JVM with a minimal runtime, because we're still going to need them to initialise the metamodel and stuff, and they depend on the CMR resolvers, which cannot be extracted and depend on the Maven farjar that is about the biggest we have
- all the work I did is very hard to keep up to date with upstream changes because it's cross-git repos so there's no move tracking and I will have to reincorporate every CMR/Model/LazyModel/Metamodel changes manually when I'm ready to commit
- I am not far from getting rid of all errors, but haven't started to test to verify that I didn't break major things with the `ModuleManager` refactoring

So I'm not quite sure how to proceed. If I don't finish this now I'm pretty sure all this work is for nothing (except the knowledge) because I won't be able to reuse my commits later.

**Comment by quintesse** [comment]
I'd say that, if possible, you make compile and work what you have so far and commit it, it is bound to be much better already than the bunch of inter-dependencies we have now. And then in the 1.3 time-frame we take another look at it. After the OBR it should hopefully be easier to split things up in several smaller pieces.

**Comment by FroMage** [comment]
Just pushed it, I will do a recap on the mailing list, but for now, know that:

- `com.redhat.ceylon.compiler.typechecker.model` (ceylon-spec) moved to `com.redhat.ceylon.model.typechecker.model`
- `com.redhat.ceylon.compiler.typechecker.context.ModuleManager` (ceylon-spec) moved to `com.redhat.ceylon.model.typechecker.util`
- `com.redhat.ceylon.compiler.typechecker.context.ProducedTypeCache` (ceylon-spec) moved to `com.redhat.ceylon.model.typechecker.context`
- `com.redhat.ceylon.compiler.typechecker.util.ProducedTypeNamePrinter` (ceylon-spec) moved to `com.redhat.ceylon.model.typechecker.util`
- `com.redhat.ceylon.compiler.java.loader` (ceylon-compiler) moved to `com.redhat.ceylon.model.loader`
- `com.redhat.ceylon.module-resolver.api{ArtifactResult,ArtifactResultType,ImportType,JDKUtils,PathFilter,Repository,RepositoryException,VisibilityType}` (ceylon-module-resolver) moved to `com.redhat.ceylon.model.cmr`
- `Decl` had some methods move to `com.redhat.ceylon.model.loader.JvmBackendUtil` and others to `com.redhat.ceylon.model.typechecker.model.ModelUtil`
- `Naming` had some methods moved to its new superclass `com.redhat.ceylon.model.loader.NamingBase`
- Annotation constructor model is now optional depending on the model loader (not for runtime for instance) and logic moved to `AnnotationLoader` (ceylon-compiler)
- `ModuleManager` is split into itself (ceylon-model) and `ModuleSourceMapper` (ceylon-spec) which takes care of error reporting
- `TypecheckerUnit` is the subtype of `Unit` that the typechecker uses for typechecking
- ceylon-model depends on ceylon-common
- ceylon.language still depends on the other modules for now, but I may be able to fix that soon

**Comment by FroMage** [comment]
So now the biggest question is LICENSE-wise, as that module contains code from `ceylon-compiler` (GPL+classpath) and other modules (`ASL`). How the hell do we resolve this?

**Comment by gavinking** [comment]
@FroMage I would be surprised if there were any code in there which we didn't write ourselves.

**Comment by FroMage** [comment]
Does that mean we can just change its license?

**Comment by gavinking** [comment]
well, yes.

**Comment by gavinking** [comment]
Unless there is code there that is written by an external contributor. But I think here we only have code where Red Hat owns the copyright.

**Comment by FroMage** [comment]
OK, let's see what I can do then.

**Comment by FroMage** [comment]
Note that I have just removed all dependencies from the language runtime to spec and compilers. Only `Main` and `CeylonToolProvider` still depend on them, and it's not trivial to fix. Which is unfortunate since I think on Android I'd use `Main` but it drags many parts of the CMR. Still, manually we should now be able to run code with a very small number of jars in the classpath (provided you initialise the metamodel manually).

**Comment by gavinking** [comment]
> Note that I have just removed all dependencies from the language runtime to spec and compilers.

That's _great_.

>  Which is unfortunate since I think on Android I'd use `Main` but it drags many parts of the CMR.

I suspect that for Android we will want a non-CMR based `Main`.

**Comment by FroMage** [comment]
Well, the thing is that the CMR currently provides us with reading of the module descriptors like `module.xml`, `module.properties`, Maven modules and OSGi.

**Comment by FroMage** [comment]
So yeah, but then we need to get those things out of the CMR, which is not trivial.

**Comment by gavinking** [comment]
@FroMage sure but it sounds like something we need to do anyway.

**Comment by FroMage** [comment]
WDYM "anyway"? It's just for that that we'd need to move it out ;) And for Maven I'm very skeptical as the biggest reason to make a lightweight runtime was to drop requirement for aether (by far biggest our jar) and guess what? You still need it (with or without CMR) to read maven module dependencies :(

**Comment by gavinking** [comment]
Well the issue is that CMR is fundamentally tied to jboss modules (isn't it?) and we need to be anticipating a future where we can run on Jigsaw or on a flat classpath or whatever.

**Comment by FroMage** [comment]
Although we could mitigate this with a tool that imports Maven modules into a Ceylon repo and adds the required `module.xml` files as it does so. This way it'd be a compile-time step to get rid of the runtime dep on aether.

**Comment by FroMage** [comment]
No, the CMR is not tied at all to jboss modules.

**Comment by FroMage** [comment]
Nor is our runtime.

**Comment by gavinking** [comment]
Oh it's not?! I did not realize that. Hah!

**Comment by FroMage** [comment]
This is where @alesj usually shows up just to brag.

**Comment by alesj** [comment]
@FroMage @gavinking yup, both CMR and Runtime have proper abstractions in place, to easily move to whatever ;-)
(brag brag ... :-)

**Comment by alesj** [comment]
Although to be honest, Runtime would need a bit fair of work, but it's doable -- specially with all the things we learned so far.
